### ISSUE TITLE
ABCI++: Decide how to handle vote extension in the block format

### ISSUE THREAD

**Issue by sergio-mena** [tendermint/tendermint#7664]
## Summary

As per title

## Problem Definition

To start with something we already know, the tendermint-signed part of the vote extensions will be breaking the light client of previous versions, so this part of the vote extensions should be postponed to 0.37 (which we know will be breaking previous versions).
The current ABCI++ code includes tendermint-signed part of the vote extensions, so I could test (thanks Anca, William for the help!) that the old light clients are indeed broken.

The question is what to do with *app_self_signed*:
* should it appear in the block?
* should it not?

It it should, then we need to find out if previous light clients would also be affected (the block hash).

But first, we need to come to an agreement on whether *app_self_signed* should appear in the block format or not.

A simple, rock-solid approach would be to simply delay vote extensions altogether to 0.37 so we can come up with a simple solution without having to care for introducing breaking changes. Any solution in 0.36 risks being more convoluted in order to keep compatibility.

## Dependencies

None

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by sergio-mena** [comment]
ABCI++ meeting on 2021-01-26:
* Meeting notes (thanks @ValarDragon !): https://hackmd.io/q7dmOAp4SZiPXTfhi6Kj8A?both
* [Slides](https://github.com/tendermint/tendermint/files/7949333/abci%2B%2B_vote_extensions.pdf) used during the meeting.

After going through the slides, the discussion focused on three possible approaches (see last [slide](https://github.com/tendermint/tendermint/files/7949333/abci%2B%2B_vote_extensions.pdf)):
1. Not including *app_signed* in the block
2. Including *app_signed* in the block, but allowing `PrepareProposal` to modify it
3. Removing *app_signed* completely, and just stay with *tendermint_signed*

By the end of the meeting, we were leaning toward solution 2 above, but we weren't 100% sure we fully understood all implications. However, something that became clear (to me) during the discussion is that this solution includes solution 1 if the App decides to delete *app_signed* at `PrepareProposal`, but with the advantage that the easy-to-grasp default behavior would be to leave it as it is in the block.

Finally, we briefly discussed the implications for the different Tendermint releases:
* The *tendermint_signed* part is clear that will only make it to 0.37 (which will anyway be changing the block structure for other reasons)
* As for *app_signed*,
  * solution 1 could be shipped in 0.36
  * solution 2 might break the block hash (see AI below); if that is the case, then target has to be 0.37, together with *tendermint_signed*
  * solution 3 is the "empty set" regarding *app_signed*

Therefore, as an AI, I (@sergio-mena ) will be testing:

* Modify the abci++ code to include *app_signed* somewhere in the block
* Test the resulting code with v0.35 light client

**Comment by sergio-mena** [comment]
After several internal discussions in the Tendermint Core team, our current approach is to release `PrepareProposal`, `ProcessProposal` and FinalizeBlock in v0.36, and release `ExtendVote` and `VerifyVoteExtension` in v0.37.

The main reason for releasing `ExtendVote` and `VerifyVoteExtension` in v0.37 is that is it likely that we need to modify the block format for those, thus breaking light clients from previous versions. Since v0.37 will anyway be changing the block format for other reasons unrelated to ABCI++, we think it is cleaner to change the block format only once in 0.37, rather than doing it both in v0.36 and v0.37.

We are making efforts at keeping the existing spec and implementation work for vote extensions in the commit history, so it will be fairly easy to restore those once the v0.36 branch separates from master.

Please let us know if you have objections w.r.t. this approach. We will close this issue in a few days.

**Comment by hdevalence** [comment]
@sergio-mena For us, the whole reason to use ABCI++ is to have access to vote extensions, so (from our point of view) this approach would be the same as delaying ABCI++ to v0.37. If there are other reasons to delay ABCI++ to v0.37, maybe that makes sense, but avoiding the need to update light clients doesn't seem compelling to us.

**Comment by cmwaters** [comment]
We had a call today to further discuss the vote extension implementation. One of the ideas that was formed over the course of the call and got a lot of consensus was rather than having these two vote extension types: `tendermint_signed_data` and `app_signed_data` where the former changes the `Commit` struct and breaks IBC, we instead implement a somewhat merge of the two. This takes the benefits of the `tendermint_signed_data` in that applications don't need to worry about key management as well as the benefits of `app_signed_data` whereby applications can modify the vote extension data in `PrepareProposal` and most importantly that it doesn't break IBC.

The summarised description of this is simply to only have `app_signed_data`  but instead of relying on the application to sign it, we reuse Tendermint's private key and sign over the returned data. An important distinction to `tendermint_signed_data` is however that instead of one there are two related but separate signatures. I will try to describe the flow in more detail. 

0. A validator has received a block that it has validated using `ProcessProposal`. It has sent out it's prevote and then received 2/3+ prevotes moving the consensus state machine to the precommit stage. 
1. In preparing the vote, it first calls `ExtendVote` and receives a blob of data (an array of bytes) from the application. It adds  the byte representation of the `Height`, `Round` and `ChainID` to the existing app data to ensure uniqueness and signs over it using the Validator's Private key until you have the following struct:
```go
type VoteExtension struct {
    Data []byte
    Height int64
    Round int32
    ChainID string
    Address string // of the validator who signed the data
}
```
2. The validator then prepares the `Precommit` vote in the same way that was done before. The contents of the bytes that is signed over does not change in any way. This ensures light client backwards compatibility for IBC.
3. The validator sends both the `VoteExtension` and the `Vote` together. Both have signatures and are verifiable. Whether they are combined into the same struct of sent over different channels is an implementation detail.
4. The validator then continues to receive votes from its peers. Along with the usual vote verification, the validator also verifies the extension by taking the address, finding the peers pubkey and verifying that the bytes and signature match. Once verified, the validator gives the blob of application data to the application via `VerifyVoteExtension`. If the application marks this as valid and the original vote is valid then the validator marks it in their tally.
5. Once 2/3+ valid `Precommit`s arrive the node executes the block. In the case that the node is the proposer in the following height, Tendermint will take all the blobs of app data from various validators and give it to the application in `PrepareProposal` in which the application can use the data to inform the `Txs` that will make it in the next block.

I have provided a guess as to what the API's would look like:

```proto
message RequestExtendVote {
    int64 height = 1;
    int32 round = 2; 
}

message ResponseExtendVote {
    repeated byte data = 1;
}

message RequestVerifyVoteExtension {
    repeated byte data = 1;
    int64 height = 2;
    int32 round = 3;
    string validator_address = 4; // we could also use validator_index if we wanted to save a bit more data
}

message ResponseVerifyVoteExtension {
    bool accept = 1;
}

message RequestPrepareProposal {
    bytes hash = 1;
    tendermint.types.Header header = 2 [(gogoproto.nullable) = false];
    repeated bytes txs = 3;
    LastCommitInfo last_commit_info = 4 [(gogoproto.nullable) = false];
    repeated Evidence byzantine_validators = 5 [(gogoproto.nullable) = false];
    repeated VoteExtension vote_extensions = 6;
    int64 max_tx_bytes = 7;
}

message VoteExtension {
   repeated bytes data = 1;
   string validator_address = 2;
}
```

**Comment by sergio-mena** [comment]
Thanks @cmwaters for the detailed proposal. A priori I don't see any problem with that. Only detail, as I mentioned in Slack, would be to add the block ID as a parameter in `RequestExtendVote` as, conceptually, a vote extension refers to a proposal. The same probably applies to `RequestVerifyVoteExtension`.
I've moved this issue back to "in progress". I've also moved issue #7655 to the top. I'll be updating the spec with the results of last night's meeting, that you summarized in your comment above. The PR will be linked to this issue.
